From 926a21a0207965949c33b10d9fad4238da0a943f Mon Sep 17 00:00:00 2001
From: Kenton Varda <kenton@cloudflare.com>
Date: Thu, 26 Mar 2020 19:15:48 -0500
Subject: [PATCH 21/22] Add methods to get heap and external memory sizes
 directly.

`GetHeapStatistics()` exists for this, but also collects a lot of other info and apparently performs non-trivial computation. These new accessors are intended to be very fast.

Signed-off-by: James M Snell <jsnell@cloudflare.com>
---
 include/v8-isolate.h | 10 ++++++++++
 src/api/api.cc       |  8 ++++++++
 2 files changed, 18 insertions(+)

diff --git a/include/v8-isolate.h b/include/v8-isolate.h
index 09373d56f62..750e67b7726 100644
--- a/include/v8-isolate.h
+++ b/include/v8-isolate.h
@@ -1056,6 +1056,16 @@ class V8_EXPORT Isolate {
   V8_DEPRECATE_SOON("Use ExternalMemoryAccounter instead.")
   int64_t AdjustAmountOfExternalAllocatedMemory(int64_t change_in_bytes);
 
+  /**
+   * Gets the current amount of external memory.
+   */
+  int64_t GetExternalMemory();
+
+  /**
+   * Gets the current total size of the heap (internal memory).
+   */
+  size_t GetHeapSize();
+
   /**
    * Returns heap profiler for this isolate. Will return NULL until the isolate
    * is initialized.
diff --git a/src/api/api.cc b/src/api/api.cc
index b711f183d72..1378b64fbef 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -10334,6 +10334,14 @@ void Isolate::GetHeapStatistics(HeapStatistics* heap_statistics) {
 #endif  // V8_ENABLE_WEBASSEMBLY
 }
 
+int64_t Isolate::GetExternalMemory() {
+  return reinterpret_cast<i::Isolate*>(this)->heap()->external_memory();
+}
+
+size_t Isolate::GetHeapSize() {
+  return reinterpret_cast<i::Isolate*>(this)->heap()->CommittedMemory();
+}
+
 size_t Isolate::NumberOfHeapSpaces() {
   return i::LAST_SPACE - i::FIRST_SPACE + 1;
 }
-- 
2.43.0

